# Section 6: Working with Multiple DataFrames

In [1]:
import pandas as pd
import numpy as np

## Introducting Five New Datasets

In this section, we will be working with data describing salary information for US colleges by major (field of study) and region. Each of these data sources stands separately, and it will be our job to piece them together with the methods we will learn in this section. 

We will begin by defining local variables that point to the URLs for the data.

In [2]:
# Dataset URL Sources
eng_url = 'https://andybek.com/pandas-eng'
state_url = 'https://andybek.com/pandas-state'
party_url = 'https://andybek.com/pandas-party'
liberal_url = 'https://andybek.com/pandas-liberal'
ivies_url = 'https://andybek.com/pandas-ivies'


Let's begin by reading in the engineering school salaries. This gives a list of 19 engineering schools in the US and their starting and mid-career median salaries.

In [3]:
eng = pd.read_csv(eng_url)

In [4]:
eng.head()

School Name  ... Mid-Career Median Salary
0   Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
1      California Institute of Technology (CIT)  ...             $123,000.00 
2                           Harvey Mudd College  ...             $122,000.00 
3  Polytechnic University of New York, Brooklyn  ...             $114,000.00 
4                                  Cooper Union  ...             $114,000.00 

[5 rows x 4 columns]

Now let's read in the state school data, which describes public universities.

In [5]:
state = pd.read_csv(state_url)

In [6]:
state.head()

School Name  ... Mid-Career Median Salary
0              University of California, Berkeley  ...             $112,000.00 
1                    University of Virginia (UVA)  ...             $103,000.00 
2                        Cal Poly San Luis Obispo  ...             $101,000.00 
3  University of California at Los Angeles (UCLA)  ...             $101,000.00 
4      University of California, San Diego (UCSD)  ...             $101,000.00 

[5 rows x 4 columns]

Next, we'll do the party data, which the Wall Street Journal categorizes as "party schools" with heavy alcohol use.

In [7]:
party = pd.read_csv(party_url)

In [8]:
party.head()

School Name  ... Mid-Career Median Salary
0  University of Illinois at Urbana-Champaign (UIUC)  ...              $96,100.00 
1               University of Maryland, College Park  ...              $95,000.00 
2     University of California, Santa Barbara (UCSB)  ...              $95,000.00 
3                  University of Texas (UT) - Austin  ...              $93,900.00 
4      State University of New York (SUNY) at Albany  ...              $92,200.00 

[5 rows x 4 columns]

Next up we have liberal arts schools, which emphasize rational thinking and first principles over technical knowledge

In [9]:
liberal = pd.read_csv(liberal_url)

In [10]:
liberal.head()

School Name  ... Mid-Career Median Salary
0  Bucknell University  ...             $110,000.00 
1   Colgate University  ...             $108,000.00 
2      Amherst College  ...             $107,000.00 
3    Lafayette College  ...             $107,000.00 
4      Bowdoin College  ...             $107,000.00 

[5 rows x 4 columns]

Last but not least, let's take a look at the Ivy League schools - a group of highly prestigious and highly selective schools in the American northeast. .

In [11]:
ivies = pd.read_csv(ivies_url)

In [12]:
ivies.head()

School Name  ... Mid-Career Median Salary
0           Dartmouth College  ...             $134,000.00 
1        Princeton University  ...             $131,000.00 
2             Yale University  ...             $126,000.00 
3          Harvard University  ...             $124,000.00 
4  University of Pennsylvania  ...             $120,000.00 

[5 rows x 4 columns]

## Concatenating DataFrames

In the previous section, we read in 5 dataframes of similar structure but fragmented. Let's check the shape of each dataframe to confirm that they are shaped similarly.

In [13]:
dfs = [state, eng, liberal, ivies, party]

In [14]:
for df in dfs:
  print(df.shape)

(175, 4)
(19, 4)
(47, 4)
(8, 4)
(20, 4)


So we have 5 two-dimensional dataframes, each with a different number of schools but the same four columns. 

As a first step, let's put them all together, a process called **concatenation**. Suppose we want to concatenate the Ivies and engineering schools. To do this, we can run the `pandas.concat()` function, passing in a list of the dataframes that we want to concatenate..
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html

In [15]:
pd.concat([ivies, eng])

School Name  ... Mid-Career Median Salary
0                                   Dartmouth College  ...             $134,000.00 
1                                Princeton University  ...             $131,000.00 
2                                     Yale University  ...             $126,000.00 
3                                  Harvard University  ...             $124,000.00 
4                          University of Pennsylvania  ...             $120,000.00 
5                                  Cornell University  ...             $110,000.00 
6                                    Brown University  ...             $109,000.00 
7                                 Columbia University  ...             $107,000.00 
0         Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
1            California Institute of Technology (CIT)  ...             $123,000.00 
2                                 Harvey Mudd College  ...             $122,000.00 
3        Polytechnic University of New York, Brooklyn  ...             $114,000.00 
4                                        Cooper Union  ...             $114,000.00 
5               Worcester Polytechnic Institute (WPI)  ...             $114,000.00 
6                    Carnegie Mellon University (CMU)  ...             $111,000.00 
7              Rensselaer Polytechnic Institute (RPI)  ...             $110,000.00 
8                     Georgia Institute of Technology  ...             $106,000.00 
9                            Colorado School of Mines  ...             $106,000.00 
10                    Stevens Institute of Technology  ...             $105,000.00 
11             Illinois Institute of Technology (IIT)  ...              $97,800.00 
12                  Wentworth Institute of Technology  ...              $96,700.00 
13  Virginia Polytechnic Institute and State Unive...  ...              $95,400.00 
14          South Dakota School of Mines & Technology  ...              $93,400.00 
15  New Mexico Institute of Mining and Technology ...  ...              $93,400.00 
16            Rochester Institute of Technology (RIT)  ...              $84,600.00 
17        Embry-Riddle Aeronautical University (ERAU)  ...              $80,700.00 
18                 Tennessee Technological University  ...              $80,000.00 

[27 rows x 4 columns]

Let's check the shape of the concatenated dataframe

In [16]:
pd.concat([ivies, eng]).shape

(27, 4)

Let's try concatenating all of these dataframes!

In [17]:
pd.concat(dfs)

School Name  ... Mid-Career Median Salary
0               University of California, Berkeley  ...             $112,000.00 
1                     University of Virginia (UVA)  ...             $103,000.00 
2                         Cal Poly San Luis Obispo  ...             $101,000.00 
3   University of California at Los Angeles (UCLA)  ...             $101,000.00 
4       University of California, San Diego (UCSD)  ...             $101,000.00 
..                                             ...  ...                      ...
15               University of New Hampshire (UNH)  ...              $78,300.00 
16                  West Virginia University (WVU)  ...              $78,100.00 
17                         University of Tennessee  ...              $74,600.00 
18                                 Ohio University  ...              $73,400.00 
19                  Florida State University (FSU)  ...              $73,000.00 

[269 rows x 4 columns]

We've now concatenated all of the dataframes into one master frame!



In [18]:
pd.concat(dfs).head()

School Name  ... Mid-Career Median Salary
0              University of California, Berkeley  ...             $112,000.00 
1                    University of Virginia (UVA)  ...             $103,000.00 
2                        Cal Poly San Luis Obispo  ...             $101,000.00 
3  University of California at Los Angeles (UCLA)  ...             $101,000.00 
4      University of California, San Diego (UCSD)  ...             $101,000.00 

[5 rows x 4 columns]

But there's a problem with this - these datasets are NOT mutually exclusive. Some schools are in more than one dataframe - for example, most party schools are also state schools.

In [19]:
pd.concat(dfs)["School Name"].value_counts()

University of Florida (UF)                        2
West Virginia University (WVU)                    2
Florida State University (FSU)                    2
Louisiana State University (LSU)                  2
University of Iowa (UI)                           2
                                                 ..
Boise State University (BSU)                      1
University of Delaware                            1
Colgate University                                1
University of Southern Maine                      1
California State University, Northridge (CSUN)    1
Name: School Name, Length: 249, dtype: int64

To see this more clearly, let's use the `set()` and `difference()` functions to compare the differences between the party and state school name lists. Remember that 
* sets do not allow for duplicate values. Thus, we will get a unique list when we call the `set()` function.
* the difference function looks for any valus that are in one collection but not the other 

In [20]:
set(party["School Name"]).difference(state['School Name'])

{'Randolph-Macon College'}

And thus we see that Randolph-Macon College is the only university that is not both a party school and a state school. For those who care, Randolph-Macon College is a private party school. It's also a liberal arts school.

In [21]:
'Randolph-Macon College' in liberal["School Name"].values

True

Thus, all party schools are either state schools, liberal arts schools, or both. 

Just for kicks, let's see if any engineering schools are also party schools using the `intersection()` set method.

In [22]:
set(eng["School Name"]).intersection(party['School Name'])

set()

Guess not.

Another way to check for duplication is to use the `duplicated()` method on the concatenated dataframe. Remember this will generate a boolean array, which we can use to make a selection.

In [23]:
pd.concat(dfs).duplicated(subset = ['School Name'], keep = 'first')

0     False
1     False
2     False
3     False
4     False
      ...  
15     True
16     True
17     True
18     True
19     True
Length: 269, dtype: bool

In [24]:
pd.concat(dfs)[pd.concat(dfs).duplicated(subset = ['School Name'], keep = 'first')]

School Name  ... Mid-Career Median Salary
0   University of Illinois at Urbana-Champaign (UIUC)  ...              $96,100.00 
1                University of Maryland, College Park  ...              $95,000.00 
2      University of California, Santa Barbara (UCSB)  ...              $95,000.00 
3                   University of Texas (UT) - Austin  ...              $93,900.00 
4       State University of New York (SUNY) at Albany  ...              $92,200.00 
5                          University of Florida (UF)  ...              $87,900.00 
6                    Louisiana State University (LSU)  ...              $87,800.00 
7                         University of Georgia (UGA)  ...              $86,000.00 
8                 Pennsylvania State University (PSU)  ...              $85,700.00 
9                      Arizona State University (ASU)  ...              $84,100.00 
10               Indiana University (IU), Bloomington  ...              $84,000.00 
11                            University of Iowa (UI)  ...              $83,900.00 
12                             Randolph-Macon College  ...              $83,600.00 
13                  University of Alabama, Tuscaloosa  ...              $81,400.00 
14                          University of Mississippi  ...              $79,700.00 
15                  University of New Hampshire (UNH)  ...              $78,300.00 
16                     West Virginia University (WVU)  ...              $78,100.00 
17                            University of Tennessee  ...              $74,600.00 
18                                    Ohio University  ...              $73,400.00 
19                     Florida State University (FSU)  ...              $73,000.00 

[20 rows x 4 columns]

So check that out, there are 20 schools that are duplicates in our dataset, and they all happen to be party schools. Because we know that our *party* dataframe has 20 rows, this means that the entire *party* dataframe consists of schools that show up in at least one other constituent dataframe. 

How can we fix this? One way is to simply remove the party schools from our list of dataframes that we use for the concatenation. 

A more programmatic way of doing this is to use the `drop_duplicates()` method, subsetting by school name.

In [25]:
pd.concat(dfs).drop_duplicates(subset = "School Name")

School Name  ... Mid-Career Median Salary
0               University of California, Berkeley  ...             $112,000.00 
1                     University of Virginia (UVA)  ...             $103,000.00 
2                         Cal Poly San Luis Obispo  ...             $101,000.00 
3   University of California at Los Angeles (UCLA)  ...             $101,000.00 
4       University of California, San Diego (UCSD)  ...             $101,000.00 
..                                             ...  ...                      ...
3                               Harvard University  ...             $124,000.00 
4                       University of Pennsylvania  ...             $120,000.00 
5                               Cornell University  ...             $110,000.00 
6                                 Brown University  ...             $109,000.00 
7                              Columbia University  ...             $107,000.00 

[249 rows x 4 columns]

Note that we've now lost the information of whether the a particular school is a party school (in addition to being a state or liberal arts school).

Let's lock this into memory by assigning to a variable.

In [26]:
schools = pd.concat(dfs).drop_duplicates(subset = "School Name")

As a side activity, let's try to remove all party schools from the concatenated data frame completely. This would allow us to do comparisons like, for example, comparing starting salaries at party schools versus non-party schools. Remember that we can't simply remove the party schools from the dataframe, because all of those schools are also present as *state* or *liberal arts* schools which are in the dataframe. We have to remove any school that is a party school completely.

To start, let's use list comprehension to create a new boolean mask that indicates whether a school **is not** in the list of party schools. This will be tested for every school in the concatenated school dataframe.

In [27]:
is_not_party_school = [school not in party["School Name"].values for school in schools["School Name"].values]

We can now use selection to grab the schools from the concatenated dataframe that are NOT party schools!

In [28]:
schools[is_not_party_school]

School Name  ... Mid-Career Median Salary
0               University of California, Berkeley  ...             $112,000.00 
1                     University of Virginia (UVA)  ...             $103,000.00 
2                         Cal Poly San Luis Obispo  ...             $101,000.00 
3   University of California at Los Angeles (UCLA)  ...             $101,000.00 
4       University of California, San Diego (UCSD)  ...             $101,000.00 
..                                             ...  ...                      ...
3                               Harvard University  ...             $124,000.00 
4                       University of Pennsylvania  ...             $120,000.00 
5                               Cornell University  ...             $110,000.00 
6                                 Brown University  ...             $109,000.00 
7                              Columbia University  ...             $107,000.00 

[229 rows x 4 columns]

## The Duplicated Index Issue

In the previous lecture, we concatenated 5 dataframes together and removed the party school. Let's look at this dataframe.


In [29]:
schools

School Name  ... Mid-Career Median Salary
0               University of California, Berkeley  ...             $112,000.00 
1                     University of Virginia (UVA)  ...             $103,000.00 
2                         Cal Poly San Luis Obispo  ...             $101,000.00 
3   University of California at Los Angeles (UCLA)  ...             $101,000.00 
4       University of California, San Diego (UCSD)  ...             $101,000.00 
..                                             ...  ...                      ...
3                               Harvard University  ...             $124,000.00 
4                       University of Pennsylvania  ...             $120,000.00 
5                               Cornell University  ...             $110,000.00 
6                                 Brown University  ...             $109,000.00 
7                              Columbia University  ...             $107,000.00 

[249 rows x 4 columns]

See the problem here? Our index for the concatenated dataframe contains duplicates indices. As an example, let's grab all rows with an index of `0`.

In [30]:
schools.loc[0]

School Name  ... Mid-Career Median Salary
0           University of California, Berkeley  ...             $112,000.00 
0  Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
0                          Bucknell University  ...             $110,000.00 
0                            Dartmouth College  ...             $134,000.00 

[4 rows x 4 columns]

This happens because the component dataframes all have indices that start at 0 and go until the end. The reason this happens is that **`pd.concat()` does not discard the original index of the dataframes being concatenated**.

To illustrate further, let's take a look at the indices using the `duplicated()` method.


In [31]:
schools.index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

As you can see, the indices of the later dataframes overlap completely with the indices of the first dataframe.

It is technically okay to have duplicated indices in Pandas dataframes. However, you will give up a lot of functionality by doing so. For example, slices by index will not work.


In [32]:
## This does NOT work with duplicated indices
# schools.loc[0:2]

For this and other reasons, it is usually best practice to have at least one common index with unique values across all records. Can we fix this for our concatenated dataframe? Yes we can!

The first method we can use is `reset_index()`. This method restores a 0-based range index in the dataframe while pushing the old index into a regular column in the dataframe.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html

In [33]:
schools.reset_index()

index  ... Mid-Career Median Salary
0        0  ...             $112,000.00 
1        1  ...             $103,000.00 
2        2  ...             $101,000.00 
3        3  ...             $101,000.00 
4        4  ...             $101,000.00 
..     ...  ...                      ...
244      3  ...             $124,000.00 
245      4  ...             $120,000.00 
246      5  ...             $110,000.00 
247      6  ...             $109,000.00 
248      7  ...             $107,000.00 

[249 rows x 5 columns]

Oftentimes however there is no use for the old index. So we can remove it entirely be setting the `drop` parameter to `True`. Let's also perform the method in place to modify the underlying dataframe.

In [34]:
schools.reset_index(drop = True, inplace=True)

In [35]:
schools

School Name  ... Mid-Career Median Salary
0                University of California, Berkeley  ...             $112,000.00 
1                      University of Virginia (UVA)  ...             $103,000.00 
2                          Cal Poly San Luis Obispo  ...             $101,000.00 
3    University of California at Los Angeles (UCLA)  ...             $101,000.00 
4        University of California, San Diego (UCSD)  ...             $101,000.00 
..                                              ...  ...                      ...
244                              Harvard University  ...             $124,000.00 
245                      University of Pennsylvania  ...             $120,000.00 
246                              Cornell University  ...             $110,000.00 
247                                Brown University  ...             $109,000.00 
248                             Columbia University  ...             $107,000.00 

[249 rows x 4 columns]

Let's make sure the duplicated indices are gone.

In [36]:
schools.index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

Beautiful. We now know that our index is unique.

An even more elegant approach is to **discard the old indices as we do the concatenation**. We can do this by modifying the `concat()` method by using the `ignore_index` parameter. This essentially tells Pandas to ignore the indices of the dataframes being concatenated. Instead, Pandas generates a new index for a new dataframe that is being produced.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html



In [37]:
pd.concat(dfs, ignore_index=True).drop_duplicates(subset = "School Name")

School Name  ... Mid-Career Median Salary
0                University of California, Berkeley  ...             $112,000.00 
1                      University of Virginia (UVA)  ...             $103,000.00 
2                          Cal Poly San Luis Obispo  ...             $101,000.00 
3    University of California at Los Angeles (UCLA)  ...             $101,000.00 
4        University of California, San Diego (UCSD)  ...             $101,000.00 
..                                              ...  ...                      ...
244                              Harvard University  ...             $124,000.00 
245                      University of Pennsylvania  ...             $120,000.00 
246                              Cornell University  ...             $110,000.00 
247                                Brown University  ...             $109,000.00 
248                             Columbia University  ...             $107,000.00 

[249 rows x 4 columns]

As a sanity check, let's check the index for duplicates.

In [38]:
pd.concat(dfs, ignore_index=True).drop_duplicates(subset = "School Name").index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

Awesome, no duplicates!

## Enforcing Unique Indices

We previously explored ways to remove the old indices of dataframes when concatenating them. But what if those indices contained useful information and we wanted to keep them, yet still created a concatenated dataframe with unique indices? Turns out we can do that. 

To illustrate, let's assume our original dataframes have the school name as the indices.

In [39]:
ivies2 = ivies.set_index('School Name')

In [40]:
ivies2

School Type  ... Mid-Career Median Salary
School Name                             ...                         
Dartmouth College           Ivy League  ...             $134,000.00 
Princeton University        Ivy League  ...             $131,000.00 
Yale University             Ivy League  ...             $126,000.00 
Harvard University          Ivy League  ...             $124,000.00 
University of Pennsylvania  Ivy League  ...             $120,000.00 
Cornell University          Ivy League  ...             $110,000.00 
Brown University            Ivy League  ...             $109,000.00 
Columbia University         Ivy League  ...             $107,000.00 

[8 rows x 3 columns]

Let's do the same for the engineering schools.

In [41]:
eng2 = eng.set_index("School Name")

In [42]:
eng2.head()

School Type  ... Mid-Career Median Salary
School Name                                                ...                         
Massachusetts Institute of Technology (MIT)   Engineering  ...             $126,000.00 
California Institute of Technology (CIT)      Engineering  ...             $123,000.00 
Harvey Mudd College                           Engineering  ...             $122,000.00 
Polytechnic University of New York, Brooklyn  Engineering  ...             $114,000.00 
Cooper Union                                  Engineering  ...             $114,000.00 

[5 rows x 3 columns]

Now, if we were to merge the two dataframes and set `ignore_index` to `True`, we would be losing important information - namely, the school names.

In [43]:
pd.concat([ivies2, eng2], ignore_index = True)

School Type Starting Median Salary Mid-Career Median Salary
0    Ivy League            $58,000.00              $134,000.00 
1    Ivy League            $66,500.00              $131,000.00 
2    Ivy League            $59,100.00              $126,000.00 
3    Ivy League            $63,400.00              $124,000.00 
4    Ivy League            $60,900.00              $120,000.00 
5    Ivy League            $60,300.00              $110,000.00 
6    Ivy League            $56,200.00              $109,000.00 
7    Ivy League            $59,400.00              $107,000.00 
8   Engineering            $72,200.00              $126,000.00 
9   Engineering            $75,500.00              $123,000.00 
10  Engineering            $71,800.00              $122,000.00 
11  Engineering            $62,400.00              $114,000.00 
12  Engineering            $62,200.00              $114,000.00 
13  Engineering            $61,000.00              $114,000.00 
14  Engineering            $61,800.00              $111,000.00 
15  Engineering            $61,100.00              $110,000.00 
16  Engineering            $58,300.00              $106,000.00 
17  Engineering            $58,100.00              $106,000.00 
18  Engineering            $60,600.00              $105,000.00 
19  Engineering            $56,000.00               $97,800.00 
20  Engineering            $53,000.00               $96,700.00 
21  Engineering            $53,500.00               $95,400.00 
22  Engineering            $55,800.00               $93,400.00 
23  Engineering            $51,000.00               $93,400.00 
24  Engineering            $48,900.00               $84,600.00 
25  Engineering            $52,700.00               $80,700.00 
26  Engineering            $46,200.00               $80,000.00

That is NOT very useful. So, how do we preserve the indices while enforcing uniqueness? 

Let's start by concatenating with the original indices.

In [44]:
pd.concat([ivies2, eng2])

School Type  ... Mid-Career Median Salary
School Name                                                      ...                         
Dartmouth College                                    Ivy League  ...             $134,000.00 
Princeton University                                 Ivy League  ...             $131,000.00 
Yale University                                      Ivy League  ...             $126,000.00 
Harvard University                                   Ivy League  ...             $124,000.00 
University of Pennsylvania                           Ivy League  ...             $120,000.00 
Cornell University                                   Ivy League  ...             $110,000.00 
Brown University                                     Ivy League  ...             $109,000.00 
Columbia University                                  Ivy League  ...             $107,000.00 
Massachusetts Institute of Technology (MIT)         Engineering  ...             $126,000.00 
California Institute of Technology (CIT)            Engineering  ...             $123,000.00 
Harvey Mudd College                                 Engineering  ...             $122,000.00 
Polytechnic University of New York, Brooklyn        Engineering  ...             $114,000.00 
Cooper Union                                        Engineering  ...             $114,000.00 
Worcester Polytechnic Institute (WPI)               Engineering  ...             $114,000.00 
Carnegie Mellon University (CMU)                    Engineering  ...             $111,000.00 
Rensselaer Polytechnic Institute (RPI)              Engineering  ...             $110,000.00 
Georgia Institute of Technology                     Engineering  ...             $106,000.00 
Colorado School of Mines                            Engineering  ...             $106,000.00 
Stevens Institute of Technology                     Engineering  ...             $105,000.00 
Illinois Institute of Technology (IIT)              Engineering  ...              $97,800.00 
Wentworth Institute of Technology                   Engineering  ...              $96,700.00 
Virginia Polytechnic Institute and State Univer...  Engineering  ...              $95,400.00 
South Dakota School of Mines & Technology           Engineering  ...              $93,400.00 
New Mexico Institute of Mining and Technology (...  Engineering  ...              $93,400.00 
Rochester Institute of Technology (RIT)             Engineering  ...              $84,600.00 
Embry-Riddle Aeronautical University (ERAU)         Engineering  ...              $80,700.00 
Tennessee Technological University                  Engineering  ...              $80,000.00 

[27 rows x 3 columns]

That's better. But how do we *ensure* that the none of the indices are duplicated (it turns out they are not in this example, but let's pretend that they could be)? We can do that by using the `verify_integrity` parameter within the `pd.concat()` method. This parameter performs a unique index check for us, in that Pandas will throw an error if it detects a duplicated index.

In [45]:
pd.concat([ivies2, eng2], verify_integrity=True)

School Type  ... Mid-Career Median Salary
School Name                                                      ...                         
Dartmouth College                                    Ivy League  ...             $134,000.00 
Princeton University                                 Ivy League  ...             $131,000.00 
Yale University                                      Ivy League  ...             $126,000.00 
Harvard University                                   Ivy League  ...             $124,000.00 
University of Pennsylvania                           Ivy League  ...             $120,000.00 
Cornell University                                   Ivy League  ...             $110,000.00 
Brown University                                     Ivy League  ...             $109,000.00 
Columbia University                                  Ivy League  ...             $107,000.00 
Massachusetts Institute of Technology (MIT)         Engineering  ...             $126,000.00 
California Institute of Technology (CIT)            Engineering  ...             $123,000.00 
Harvey Mudd College                                 Engineering  ...             $122,000.00 
Polytechnic University of New York, Brooklyn        Engineering  ...             $114,000.00 
Cooper Union                                        Engineering  ...             $114,000.00 
Worcester Polytechnic Institute (WPI)               Engineering  ...             $114,000.00 
Carnegie Mellon University (CMU)                    Engineering  ...             $111,000.00 
Rensselaer Polytechnic Institute (RPI)              Engineering  ...             $110,000.00 
Georgia Institute of Technology                     Engineering  ...             $106,000.00 
Colorado School of Mines                            Engineering  ...             $106,000.00 
Stevens Institute of Technology                     Engineering  ...             $105,000.00 
Illinois Institute of Technology (IIT)              Engineering  ...              $97,800.00 
Wentworth Institute of Technology                   Engineering  ...              $96,700.00 
Virginia Polytechnic Institute and State Univer...  Engineering  ...              $95,400.00 
South Dakota School of Mines & Technology           Engineering  ...              $93,400.00 
New Mexico Institute of Mining and Technology (...  Engineering  ...              $93,400.00 
Rochester Institute of Technology (RIT)             Engineering  ...              $84,600.00 
Embry-Riddle Aeronautical University (ERAU)         Engineering  ...              $80,700.00 
Tennessee Technological University                  Engineering  ...              $80,000.00 

[27 rows x 3 columns]

Turns out our dataframes had unique indices and thus no error was thrown. Let's mess around with our data to illustrate the power of `verify_integrity`.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html

Let's pick a random sample from the *eng2* dataframe and assign it to a new variable.

In [46]:
random_eng_school = eng2.sample()

In [47]:
random_eng_school

School Type  ... Mid-Career Median Salary
School Name                                             ...                         
South Dakota School of Mines & Technology  Engineering  ...              $93,400.00 

[1 rows x 3 columns]

Now let's add this school to the *ivies2* dataframe.

In [48]:
ivies2 = ivies2.append(random_eng_school)

In [49]:
ivies2

School Type  ... Mid-Career Median Salary
School Name                                             ...                         
Dartmouth College                           Ivy League  ...             $134,000.00 
Princeton University                        Ivy League  ...             $131,000.00 
Yale University                             Ivy League  ...             $126,000.00 
Harvard University                          Ivy League  ...             $124,000.00 
University of Pennsylvania                  Ivy League  ...             $120,000.00 
Cornell University                          Ivy League  ...             $110,000.00 
Brown University                            Ivy League  ...             $109,000.00 
Columbia University                         Ivy League  ...             $107,000.00 
South Dakota School of Mines & Technology  Engineering  ...              $93,400.00 

[9 rows x 3 columns]

We now have the Worcester Polytechnic Institute which is part of both *ivies2* and *eng2*. Now let's try to concatenated them with integrity verification. It throws an error!

In [50]:
## This will throw an error!
# pd.concat([ivies2, eng2], verify_integrity=True)

To get around this and allow us to concatenate the dataframes with shared indices, we have a few options.
1. We could choose a different index for the two dataframes that *do* have unique values. The instructor recommends this approach.
2. We can turn of `verify_integrity` and then deal with the duplicates afterward, either by dropping duplicates or resetting the index.

One final note is that you can also concatenate along axis 1 (the column axis).

## Creating Multiple Indices with `concat()`

We previously looked at how to replace the index of a concatenated dataframe with a new zero-based index using the `ignore_index` parameter of `concat()`. But what if we wanted to add another level of indexing to easily partition or identify the constituent dataframes within the structure of the concatenated dataframe. We'll explore this more in the next section, but here's a sneak peak.

Say we're combining the ivies and the engineering schools again, but we want to be able to specify within the index the origin of the data. Recall that when using the vanilla `concat()` function, we'll get repeating indices.


In [51]:
pd.concat([ivies, eng])

School Name  ... Mid-Career Median Salary
0                                   Dartmouth College  ...             $134,000.00 
1                                Princeton University  ...             $131,000.00 
2                                     Yale University  ...             $126,000.00 
3                                  Harvard University  ...             $124,000.00 
4                          University of Pennsylvania  ...             $120,000.00 
5                                  Cornell University  ...             $110,000.00 
6                                    Brown University  ...             $109,000.00 
7                                 Columbia University  ...             $107,000.00 
0         Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
1            California Institute of Technology (CIT)  ...             $123,000.00 
2                                 Harvey Mudd College  ...             $122,000.00 
3        Polytechnic University of New York, Brooklyn  ...             $114,000.00 
4                                        Cooper Union  ...             $114,000.00 
5               Worcester Polytechnic Institute (WPI)  ...             $114,000.00 
6                    Carnegie Mellon University (CMU)  ...             $111,000.00 
7              Rensselaer Polytechnic Institute (RPI)  ...             $110,000.00 
8                     Georgia Institute of Technology  ...             $106,000.00 
9                            Colorado School of Mines  ...             $106,000.00 
10                    Stevens Institute of Technology  ...             $105,000.00 
11             Illinois Institute of Technology (IIT)  ...              $97,800.00 
12                  Wentworth Institute of Technology  ...              $96,700.00 
13  Virginia Polytechnic Institute and State Unive...  ...              $95,400.00 
14          South Dakota School of Mines & Technology  ...              $93,400.00 
15  New Mexico Institute of Mining and Technology ...  ...              $93,400.00 
16            Rochester Institute of Technology (RIT)  ...              $84,600.00 
17        Embry-Riddle Aeronautical University (ERAU)  ...              $80,700.00 
18                 Tennessee Technological University  ...              $80,000.00 

[27 rows x 4 columns]

The `pd.concat()` function has a `keys` parameter that constructs a hierarchical index using the passed into keys as the outermost level. The keys are applied to the data respectively to according to how the constituent dataframes were passed into `pd.concat()`. 


In [52]:
new_df = pd.concat([ivies, eng], keys = ["ivyleague_schools", "engineering_schools"])

In [53]:
new_df

School Name  ... Mid-Career Median Salary
ivyleague_schools   0                                   Dartmouth College  ...             $134,000.00 
                    1                                Princeton University  ...             $131,000.00 
                    2                                     Yale University  ...             $126,000.00 
                    3                                  Harvard University  ...             $124,000.00 
                    4                          University of Pennsylvania  ...             $120,000.00 
                    5                                  Cornell University  ...             $110,000.00 
                    6                                    Brown University  ...             $109,000.00 
                    7                                 Columbia University  ...             $107,000.00 
engineering_schools 0         Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
                    1            California Institute of Technology (CIT)  ...             $123,000.00 
                    2                                 Harvey Mudd College  ...             $122,000.00 
                    3        Polytechnic University of New York, Brooklyn  ...             $114,000.00 
                    4                                        Cooper Union  ...             $114,000.00 
                    5               Worcester Polytechnic Institute (WPI)  ...             $114,000.00 
                    6                    Carnegie Mellon University (CMU)  ...             $111,000.00 
                    7              Rensselaer Polytechnic Institute (RPI)  ...             $110,000.00 
                    8                     Georgia Institute of Technology  ...             $106,000.00 
                    9                            Colorado School of Mines  ...             $106,000.00 
                    10                    Stevens Institute of Technology  ...             $105,000.00 
                    11             Illinois Institute of Technology (IIT)  ...              $97,800.00 
                    12                  Wentworth Institute of Technology  ...              $96,700.00 
                    13  Virginia Polytechnic Institute and State Unive...  ...              $95,400.00 
                    14          South Dakota School of Mines & Technology  ...              $93,400.00 
                    15  New Mexico Institute of Mining and Technology ...  ...              $93,400.00 
                    16            Rochester Institute of Technology (RIT)  ...              $84,600.00 
                    17        Embry-Riddle Aeronautical University (ERAU)  ...              $80,700.00 
                    18                 Tennessee Technological University  ...              $80,000.00 

[27 rows x 4 columns]

Our new dataframe contains an index with two levels. The first level is the numerical index from the original dataframes. Then the second (outer) level contains the labels (keys) that we specified. And if we check the indices, we see that each index is a tuple.

In [54]:
new_df.index

MultiIndex([(  'ivyleague_schools',  0),
            (  'ivyleague_schools',  1),
            (  'ivyleague_schools',  2),
            (  'ivyleague_schools',  3),
            (  'ivyleague_schools',  4),
            (  'ivyleague_schools',  5),
            (  'ivyleague_schools',  6),
            (  'ivyleague_schools',  7),
            ('engineering_schools',  0),
            ('engineering_schools',  1),
            ('engineering_schools',  2),
            ('engineering_schools',  3),
            ('engineering_schools',  4),
            ('engineering_schools',  5),
            ('engineering_schools',  6),
            ('engineering_schools',  7),
            ('engineering_schools',  8),
            ('engineering_schools',  9),
            ('engineering_schools', 10),
            ('engineering_schools', 11),
            ('engineering_schools', 12),
            ('engineering_schools', 13),
            ('engineering_schools', 14),
            ('engineering_schools', 15),
            ('en

This means that we cannot use methods like `loc[]` in the traditional sense. Instead, we have to identify the entire tupled index.

In [55]:
## This does not work with multilevel indices.
# new_df.loc[3]

In [56]:
## This will work
new_df.loc[('ivyleague_schools', 3)]

School Name                 Harvard University
School Type                         Ivy League
Starting Median Salary             $63,400.00 
Mid-Career Median Salary          $124,000.00 
Name: (ivyleague_schools, 3), dtype: object

We can also avoid this issue by using position-based indexing instead of label-based indexing. Of course, this means you must know where the data you want is located within the dataframe.

In [57]:
new_df.iloc[3]

School Name                 Harvard University
School Type                         Ivy League
Starting Median Salary             $63,400.00 
Mid-Career Median Salary          $124,000.00 
Name: (ivyleague_schools, 3), dtype: object

## Column Axis Concatenation

We previously concatenated only by rows, or along the index axis (as opposed to the column axis). But we can change this behavior and concatenate along the column axis.

To do this, we simply alter the `pd.concat()` method by changing the `axis` parameter from 0 to 1. When might this be useful?

Suppose we want to reflect, side-by-side, the top 5 Ivy League and engineering schools that produce the highest earning graduates. One way to approach this is to produce sorted copies of the dataframes. Let's grab the top 5 starting salaries from each dataframe sorted in descending order, drop the original indices, and reset the index.

In [58]:
ivies3 = ivies.sort_values(by = "Starting Median Salary", ascending=False)[0:5].reset_index(drop=True)

In [59]:
ivies3

School Name  ... Mid-Career Median Salary
0        Princeton University  ...             $131,000.00 
1          Harvard University  ...             $124,000.00 
2  University of Pennsylvania  ...             $120,000.00 
3          Cornell University  ...             $110,000.00 
4         Columbia University  ...             $107,000.00 

[5 rows x 4 columns]

In [60]:
eng3 = eng.sort_values(by = "Starting Median Salary", ascending=False)[0:5].reset_index(drop=True)

In [61]:
eng3

School Name  ... Mid-Career Median Salary
0      California Institute of Technology (CIT)  ...             $123,000.00 
1   Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
2                           Harvey Mudd College  ...             $122,000.00 
3  Polytechnic University of New York, Brooklyn  ...             $114,000.00 
4                                  Cooper Union  ...             $114,000.00 

[5 rows x 4 columns]

Finally, let's put these dataframes side-by-side in a new dataframe.

In [62]:
pd.concat([ivies3, eng3], axis = 1)

School Name  ... Mid-Career Median Salary
0        Princeton University  ...             $123,000.00 
1          Harvard University  ...             $126,000.00 
2  University of Pennsylvania  ...             $122,000.00 
3          Cornell University  ...             $114,000.00 
4         Columbia University  ...             $114,000.00 

[5 rows x 8 columns]

This looks pretty decent. For example, we can easily see that in general, the starting median salaries of top engineering schools compete well with the salaries at Ivy League schools. Interesting. Don't jump to any conclusions though - there are a lot of questions to keep in mind, including:
1. What if we considered only engineering studies at Ivy League schools? Would the starting median salary increase in that case?
2. What if we considered non-engineering students at the top engineering schools? Would the median salaries decrease?

It is possible the Ivy Leagues simply have more students that enter lower-paying, non-engineering careers, which brings down the median starting salary. It is entirely possible that engineering students at Ivy Leagues also make a good amount of money. 

A final word of note is that while this is not terrible to look at, there are better ways to visualize this type of data. This was simply to illustrate the ability of the `concat()` method to combine dataframes along the column axis.

## The `append()` Method: A Special Case of `concat()`

We previously used the `append()` method in the previous section to add new rows onto dataframes. You may have noticed that `append()` and `concat()` are similar in that regard.

For instance, we can use `append()` to add the *party* dataframe onto the *libaral* dataframe.

In [63]:
liberal.append(party)

School Name  ... Mid-Career Median Salary
0                 Bucknell University  ...             $110,000.00 
1                  Colgate University  ...             $108,000.00 
2                     Amherst College  ...             $107,000.00 
3                   Lafayette College  ...             $107,000.00 
4                     Bowdoin College  ...             $107,000.00 
..                                ...  ...                      ...
15  University of New Hampshire (UNH)  ...              $78,300.00 
16     West Virginia University (WVU)  ...              $78,100.00 
17            University of Tennessee  ...              $74,600.00 
18                    Ohio University  ...              $73,400.00 
19     Florida State University (FSU)  ...              $73,000.00 

[67 rows x 4 columns]

But equivalently, we can use concat() to do this as well.

In [64]:
pd.concat([liberal, party])

School Name  ... Mid-Career Median Salary
0                 Bucknell University  ...             $110,000.00 
1                  Colgate University  ...             $108,000.00 
2                     Amherst College  ...             $107,000.00 
3                   Lafayette College  ...             $107,000.00 
4                     Bowdoin College  ...             $107,000.00 
..                                ...  ...                      ...
15  University of New Hampshire (UNH)  ...              $78,300.00 
16     West Virginia University (WVU)  ...              $78,100.00 
17            University of Tennessee  ...              $74,600.00 
18                    Ohio University  ...              $73,400.00 
19     Florida State University (FSU)  ...              $73,000.00 

[67 rows x 4 columns]

So what's the difference? In this use case, the result is identical. However, there are differences in how they behave in different contexts and also what options you have.
* `append()` is an instance method which can only be called on existing dataframe or series objects, whereas `concat()` is a Pandas method available on the main Pandas namespace.
* `concat()` is generally more flexible in that it gives us the ability to concatenate along both the index and column axes. `append()` does not do column-wise concatenation; it's axis of operation is fixed to `0`, that is, the index axis.

## Concat() on Different Columns

So far, we've been working with dataframes that stack very nicely - they have the same columns and the same column names. Outstanding!

But sometimes we need to concatenate dataframes that don't necessarily have the same columns. For instance, let's add a STEM column to engineering schools that indicates whether or not the school specializes in Science, Technology, Engineering, or Medicine. We'll add this to a copy of *eng* so that we don't affect the original dataframe.

In [65]:
eng4 = eng.copy()

In [66]:
eng4['STEM'] = True

In [67]:
eng4.head()

School Name  ...  STEM
0   Massachusetts Institute of Technology (MIT)  ...  True
1      California Institute of Technology (CIT)  ...  True
2                           Harvey Mudd College  ...  True
3  Polytechnic University of New York, Brooklyn  ...  True
4                                  Cooper Union  ...  True

[5 rows x 5 columns]

Great, now let's try to go back and concatenate these dataframe with one of our original *ivies* dataframe and see what happens.

In [68]:
pd.concat([ivies, eng4])

School Name  ...  STEM
0                                   Dartmouth College  ...   NaN
1                                Princeton University  ...   NaN
2                                     Yale University  ...   NaN
3                                  Harvard University  ...   NaN
4                          University of Pennsylvania  ...   NaN
5                                  Cornell University  ...   NaN
6                                    Brown University  ...   NaN
7                                 Columbia University  ...   NaN
0         Massachusetts Institute of Technology (MIT)  ...  True
1            California Institute of Technology (CIT)  ...  True
2                                 Harvey Mudd College  ...  True
3        Polytechnic University of New York, Brooklyn  ...  True
4                                        Cooper Union  ...  True
5               Worcester Polytechnic Institute (WPI)  ...  True
6                    Carnegie Mellon University (CMU)  ...  True
7              Rensselaer Polytechnic Institute (RPI)  ...  True
8                     Georgia Institute of Technology  ...  True
9                            Colorado School of Mines  ...  True
10                    Stevens Institute of Technology  ...  True
11             Illinois Institute of Technology (IIT)  ...  True
12                  Wentworth Institute of Technology  ...  True
13  Virginia Polytechnic Institute and State Unive...  ...  True
14          South Dakota School of Mines & Technology  ...  True
15  New Mexico Institute of Mining and Technology ...  ...  True
16            Rochester Institute of Technology (RIT)  ...  True
17        Embry-Riddle Aeronautical University (ERAU)  ...  True
18                 Tennessee Technological University  ...  True

[27 rows x 5 columns]

As we see, the concatenated dataframe includes the STEM column, and for the Ivy League schools the value defaults to NaN. The reason for this behavior is because the `concat()` method, by default, includes all of the columns of the original dataframes. This is known as an *outer* join.
* https://radacad.com/wp-content/uploads/2015/07/joins.jpg

That said, `concat()` has a `join` parameter where we can select the type of join. For example, if we choose an *inner* join, only columns that are common to both dataframes will be in the final concatenated dataframe.

In [69]:
pd.concat([ivies, eng4], join='inner')

School Name  ... Mid-Career Median Salary
0                                   Dartmouth College  ...             $134,000.00 
1                                Princeton University  ...             $131,000.00 
2                                     Yale University  ...             $126,000.00 
3                                  Harvard University  ...             $124,000.00 
4                          University of Pennsylvania  ...             $120,000.00 
5                                  Cornell University  ...             $110,000.00 
6                                    Brown University  ...             $109,000.00 
7                                 Columbia University  ...             $107,000.00 
0         Massachusetts Institute of Technology (MIT)  ...             $126,000.00 
1            California Institute of Technology (CIT)  ...             $123,000.00 
2                                 Harvey Mudd College  ...             $122,000.00 
3        Polytechnic University of New York, Brooklyn  ...             $114,000.00 
4                                        Cooper Union  ...             $114,000.00 
5               Worcester Polytechnic Institute (WPI)  ...             $114,000.00 
6                    Carnegie Mellon University (CMU)  ...             $111,000.00 
7              Rensselaer Polytechnic Institute (RPI)  ...             $110,000.00 
8                     Georgia Institute of Technology  ...             $106,000.00 
9                            Colorado School of Mines  ...             $106,000.00 
10                    Stevens Institute of Technology  ...             $105,000.00 
11             Illinois Institute of Technology (IIT)  ...              $97,800.00 
12                  Wentworth Institute of Technology  ...              $96,700.00 
13  Virginia Polytechnic Institute and State Unive...  ...              $95,400.00 
14          South Dakota School of Mines & Technology  ...              $93,400.00 
15  New Mexico Institute of Mining and Technology ...  ...              $93,400.00 
16            Rochester Institute of Technology (RIT)  ...              $84,600.00 
17        Embry-Riddle Aeronautical University (ERAU)  ...              $80,700.00 
18                 Tennessee Technological University  ...              $80,000.00 

[27 rows x 4 columns]

In addition to having extra columns, you may also have columns that have the same type of data and intention, but have different column names. In those cases `pd.concat()` is not the best method to use. There are other techniques for dealing with this and we'll cover those later.

## Skill Challenge


#### 1. Concatenate the *liberal* and *state* schools into a new dataframe. How many unique school names are there?

Let's start by concatenating the two dataframes:

In [70]:
pd.concat([liberal, state])

School Name  ... Mid-Career Median Salary
0                    Bucknell University  ...             $110,000.00 
1                     Colgate University  ...             $108,000.00 
2                        Amherst College  ...             $107,000.00 
3                      Lafayette College  ...             $107,000.00 
4                        Bowdoin College  ...             $107,000.00 
..                                   ...  ...                      ...
170         Austin Peay State University  ...              $59,200.00 
171           Pittsburg State University  ...              $58,200.00 
172             Southern Utah University  ...              $56,500.00 
173  Montana State University - Billings  ...              $50,600.00 
174         Black Hills State University  ...              $43,900.00 

[222 rows x 4 columns]

Since we're interested in to comparing school names, let's isolate that column.

In [71]:
pd.concat([liberal, state]).loc[:, 'School Name']

0                      Bucknell University
1                       Colgate University
2                          Amherst College
3                        Lafayette College
4                          Bowdoin College
                      ...                 
170           Austin Peay State University
171             Pittsburg State University
172               Southern Utah University
173    Montana State University - Billings
174           Black Hills State University
Name: School Name, Length: 222, dtype: object

We see that this concatenated dataframe has 222 rows in it. Are any of these duplicated (that is, are any schools both state schools and liberal arts schools)?

In [72]:
pd.concat([liberal, state])['School Name'].unique()

array(['Bucknell University', 'Colgate University', 'Amherst College',
       'Lafayette College', 'Bowdoin College',
       'College of the Holy Cross', 'Occidental College',
       'Washington and Lee University', 'Swarthmore College',
       'Davidson College', 'Carleton College', 'Williams College',
       'Pomona College', 'Wesleyan University (Middletown, Connecticut)',
       'Bates College', 'Union College', 'University of Richmond',
       'Vassar College', 'Middlebury College', 'Mount Holyoke College',
       'Franklin and Marshall College', 'DePauw University',
       'St. Olaf College', 'Colby College', 'Gettysburg College',
       'Siena College', 'Smith College', 'Hamilton College',
       'Randolph-Macon College', 'Wellesley College',
       'Denison University', 'Oberlin College',
       'University of Puget Sound', 'Colorado College (CC)',
       'Reed College', 'Gustavus Adolphus College', 'Whitman College',
       'Ursinus College', 'Juniata College', 'Wittenberg Uni

It appears that there are 222 unique names, and thus there are no duplicates in this concatenated dataframe. Just to confirm, let's run a `duplicated()` and `value_counts()` on the column. 

In [73]:
pd.concat([liberal, state])['School Name'].duplicated().value_counts()

False    222
Name: School Name, dtype: int64

An alternative approach is to simply use the `nunique()` method.

In [74]:
pd.concat([liberal, state])['School Name'].nunique()

222

#### 2. Calculate the average median starting salary in the dataframe that we created in Part 1. Because the datatype for the salaries is "object", we'll need to convert it to "float".

In [75]:
pd.concat([liberal, state]).loc[:, "Starting Median Salary"]

0      $54,100.00 
1      $52,800.00 
2      $54,500.00 
3      $53,900.00 
4      $48,100.00 
          ...     
170    $37,700.00 
171    $40,400.00 
172    $41,900.00 
173    $37,900.00 
174    $35,300.00 
Name: Starting Median Salary, Length: 222, dtype: object

The presence of the commas the dollar signs makes this a bit challenging. Let's start by removing the dollar signs using the `replace()` method and regular expressions. 

In [76]:
pd.concat([liberal, state]).loc[:, "Starting Median Salary"].replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True)

0      54100.00 
1      52800.00 
2      54500.00 
3      53900.00 
4      48100.00 
         ...    
170    37700.00 
171    40400.00 
172    41900.00 
173    37900.00 
174    35300.00 
Name: Starting Median Salary, Length: 222, dtype: object

Note that an alternative approach would have been to select the dollar signs ($) and commas (,) and replaced them with an tmpty strings, which effectively removes them.

Now let's convert this all to floats. We can do this with the `replace()` function in combination with regular expressions.

In [77]:
pd.concat([liberal, state]).loc[:, "Starting Median Salary"].replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True).astype(np.float64)

0      54100.0
1      52800.0
2      54500.0
3      53900.0
4      48100.0
        ...   
170    37700.0
171    40400.0
172    41900.0
173    37900.0
174    35300.0
Name: Starting Median Salary, Length: 222, dtype: float64

Finally, let's calculate the average median starting salary.

In [78]:
pd.concat([liberal, state]).loc[:, "Starting Median Salary"].replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True).astype(np.float64).mean()

44469.36936936937

Just for fun, let's do the same calculation for the "Mid-Career Median Salary".

In [79]:
pd.concat([liberal, state]).loc[:, "Mid-Career Median Salary"].replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True).astype(np.float64).mean()

80856.3063063063

#### 3. Create a short dataframe that shows the top 3 *liberal* and *state* schools that produce the highest (mid-career) earning graduates. Show the *School Name* and *Mid-Career Median Salary* columns from each datasheet, side-by-side (i.e. horizontally). BONUS: nest the column labels within 'Liberal Arts' and 'State' Labels

We start by sorting the top 3 liberal arts and state schools by mid-career earning graduates and isolating the top 3. Remember that the values are currently in the "object" datatype, so we have to convert them to do any numerical work.

We'll start by converting both the *liberal* and *state* dataframes' dollar values from strings to floats. Then we will sort these individual dataframes by those numerical values for *Mid-Career Median Salary*, select the top 3 from each, and concatenate them together>

In [80]:
liberal_numeric = liberal.replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True)

In [81]:
liberal_numeric = liberal_numeric.astype(
    {
        "Starting Median Salary" : np.float64,
        "Mid-Career Median Salary" : np.float64
    }
)

In [82]:
liberal_numeric

School Name  ... Mid-Career Median Salary
0                             Bucknell University  ...                 110000.0
1                              Colgate University  ...                 108000.0
2                                 Amherst College  ...                 107000.0
3                               Lafayette College  ...                 107000.0
4                                 Bowdoin College  ...                 107000.0
5                       College of the Holy Cross  ...                 106000.0
6                              Occidental College  ...                 105000.0
7                   Washington and Lee University  ...                 104000.0
8                              Swarthmore College  ...                 104000.0
9                                Davidson College  ...                 104000.0
10                               Carleton College  ...                 103000.0
11                               Williams College  ...                 102000.0
12                                 Pomona College  ...                 101000.0
13  Wesleyan University (Middletown, Connecticut)  ...                  97900.0
14                                  Bates College  ...                  96500.0
15                                  Union College  ...                  95800.0
16                         University of Richmond  ...                  94600.0
17                                 Vassar College  ...                  94600.0
18                             Middlebury College  ...                  94200.0
19                          Mount Holyoke College  ...                  94100.0
20                  Franklin and Marshall College  ...                  92800.0
21                              DePauw University  ...                  88300.0
22                               St. Olaf College  ...                  86200.0
23                                  Colby College  ...                  85800.0
24                             Gettysburg College  ...                  85800.0
25                                  Siena College  ...                  85200.0
26                                  Smith College  ...                  83900.0
27                               Hamilton College  ...                  83700.0
28                         Randolph-Macon College  ...                  83600.0
29                              Wellesley College  ...                  83500.0
30                             Denison University  ...                  83500.0
31                                Oberlin College  ...                  81600.0
32                      University of Puget Sound  ...                  81500.0
33                          Colorado College (CC)  ...                  81400.0
34                                   Reed College  ...                  81100.0
35                      Gustavus Adolphus College  ...                  80600.0
36                                Whitman College  ...                  80100.0
37                                Ursinus College  ...                  80000.0
38                                Juniata College  ...                  78900.0
39                          Wittenberg University  ...                  78200.0
40                               Grinnell College  ...                  76600.0
41                               Skidmore College  ...                  74600.0
42                               Moravian College  ...                  74400.0
43                          Lewis & Clark College  ...                  72600.0
44                             Fort Lewis College  ...                  69800.0
45                         Thomas Aquinas College  ...                  67500.0
46                        Evergreen State College  ...                  63900.0

[47 rows x 4 columns]

In [83]:
liberal_numeric.sort_values(
    by = "Mid-Career Median Salary",
    ascending = False
)

School Name  ... Mid-Career Median Salary
0                             Bucknell University  ...                 110000.0
1                              Colgate University  ...                 108000.0
2                                 Amherst College  ...                 107000.0
3                               Lafayette College  ...                 107000.0
4                                 Bowdoin College  ...                 107000.0
5                       College of the Holy Cross  ...                 106000.0
6                              Occidental College  ...                 105000.0
7                   Washington and Lee University  ...                 104000.0
8                              Swarthmore College  ...                 104000.0
9                                Davidson College  ...                 104000.0
10                               Carleton College  ...                 103000.0
11                               Williams College  ...                 102000.0
12                                 Pomona College  ...                 101000.0
13  Wesleyan University (Middletown, Connecticut)  ...                  97900.0
14                                  Bates College  ...                  96500.0
15                                  Union College  ...                  95800.0
16                         University of Richmond  ...                  94600.0
17                                 Vassar College  ...                  94600.0
18                             Middlebury College  ...                  94200.0
19                          Mount Holyoke College  ...                  94100.0
20                  Franklin and Marshall College  ...                  92800.0
21                              DePauw University  ...                  88300.0
22                               St. Olaf College  ...                  86200.0
24                             Gettysburg College  ...                  85800.0
23                                  Colby College  ...                  85800.0
25                                  Siena College  ...                  85200.0
26                                  Smith College  ...                  83900.0
27                               Hamilton College  ...                  83700.0
28                         Randolph-Macon College  ...                  83600.0
29                              Wellesley College  ...                  83500.0
30                             Denison University  ...                  83500.0
31                                Oberlin College  ...                  81600.0
32                      University of Puget Sound  ...                  81500.0
33                          Colorado College (CC)  ...                  81400.0
34                                   Reed College  ...                  81100.0
35                      Gustavus Adolphus College  ...                  80600.0
36                                Whitman College  ...                  80100.0
37                                Ursinus College  ...                  80000.0
38                                Juniata College  ...                  78900.0
39                          Wittenberg University  ...                  78200.0
40                               Grinnell College  ...                  76600.0
41                               Skidmore College  ...                  74600.0
42                               Moravian College  ...                  74400.0
43                          Lewis & Clark College  ...                  72600.0
44                             Fort Lewis College  ...                  69800.0
45                         Thomas Aquinas College  ...                  67500.0
46                        Evergreen State College  ...                  63900.0

[47 rows x 4 columns]

In [84]:
state_numeric = state.replace(to_replace = '\$(\d+),*(\d+\.\d+)', value = '\\1\\2', regex = True)

In [85]:
state_numeric = state_numeric.astype(
    {
        "Starting Median Salary" : np.float64,
        "Mid-Career Median Salary" : np.float64
    }
)

In [86]:
state_numeric.sort_values(by = "Mid-Career Median Salary", ascending = False)

School Name  ... Mid-Career Median Salary
0                University of California, Berkeley  ...                 112000.0
1                      University of Virginia (UVA)  ...                 103000.0
2                          Cal Poly San Luis Obispo  ...                 101000.0
3    University of California at Los Angeles (UCLA)  ...                 101000.0
4        University of California, San Diego (UCSD)  ...                 101000.0
..                                              ...  ...                      ...
170                    Austin Peay State University  ...                  59200.0
171                      Pittsburg State University  ...                  58200.0
172                        Southern Utah University  ...                  56500.0
173             Montana State University - Billings  ...                  50600.0
174                    Black Hills State University  ...                  43900.0

[175 rows x 4 columns]

Collect the top 3 schools by *Mid-Career Median Salary* from each.

In [87]:
state_top3 = state_numeric.sort_values(by = "Mid-Career Median Salary", ascending = False).iloc[0:3]

In [88]:
state_top3

School Name  ... Mid-Career Median Salary
0  University of California, Berkeley  ...                 112000.0
1        University of Virginia (UVA)  ...                 103000.0
2            Cal Poly San Luis Obispo  ...                 101000.0

[3 rows x 4 columns]

In [89]:
liberal_top3 = liberal_numeric.sort_values(
    by = "Mid-Career Median Salary",
    ascending = False
).iloc[0:3]

In [90]:
liberal_top3

School Name  ... Mid-Career Median Salary
0  Bucknell University  ...                 110000.0
1   Colgate University  ...                 108000.0
2      Amherst College  ...                 107000.0

[3 rows x 4 columns]

Finally, concatenate them along the column axis and nest within the "Liberal Arts" and "State" labels using `pd.concat()`

In [91]:
pd.concat([state_top3, liberal_top3], keys = ['State', 'Liberal Arts'], axis = 1)

State  ...             Liberal Arts
                          School Name  ... Mid-Career Median Salary
0  University of California, Berkeley  ...                 110000.0
1        University of Virginia (UVA)  ...                 108000.0
2            Cal Poly San Luis Obispo  ...                 107000.0

[3 rows x 8 columns]

Note that the instructor video (as of 10/10/2021) has an error for Part 3 section. He forgot to convert the object values for the salaries to numeric. Didactically this is a minor detail as the section was focused on concatenating and shaping.

## The `merge()` Method

The `merge()` method allows for dataframe concatenation in a way that is very similar to SQL. This method gives us a flexible interface to join various dataframe and series options.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html

Note that `merge()` is also a dataframe method.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html

How is *merging* different from *concatenation*? 
* Think of concatenation as "gluing" datasets together. It is a structure-focused operation and does not concern itself much with the content of the data.
* On the other hand, merging combines data sets together based on the content they share. As such, it is much more flexible than concatenation. See the instructor's course notes.

With that, let's play around with `merge()`. We begin by reading in some new data describing regional information on the schools in our dataset.

In [92]:
regions_url = 'https://andybek.com/pandas-regions'

In [93]:
regions = pd.read_csv(regions_url)

In [94]:
regions.shape

(269, 2)

In [95]:
regions.head()

School Name        Region
0   Massachusetts Institute of Technology (MIT)  Northeastern
1      California Institute of Technology (CIT)    California
2                           Harvey Mudd College    California
3  Polytechnic University of New York, Brooklyn  Northeastern
4                                  Cooper Union  Northeastern

Our goal is to extend the schools dataframe by adding the regions information! Remember that the *schools* dataframe already has a column with "School Name", just like the *regions* dataframe.

In [96]:
schools.head()

School Name  ... Mid-Career Median Salary
0              University of California, Berkeley  ...             $112,000.00 
1                    University of Virginia (UVA)  ...             $103,000.00 
2                        Cal Poly San Luis Obispo  ...             $101,000.00 
3  University of California at Los Angeles (UCLA)  ...             $101,000.00 
4      University of California, San Diego (UCSD)  ...             $101,000.00 

[5 rows x 4 columns]

Let's merge that sucker.

In [97]:
pd.merge(schools, regions)

School Name  ...        Region
0                University of California, Berkeley  ...    California
1                      University of Virginia (UVA)  ...      Southern
2                          Cal Poly San Luis Obispo  ...    California
3    University of California at Los Angeles (UCLA)  ...    California
4        University of California, San Diego (UCSD)  ...    California
..                                              ...  ...           ...
264                              Harvard University  ...  Northeastern
265                      University of Pennsylvania  ...  Northeastern
266                              Cornell University  ...  Northeastern
267                                Brown University  ...  Northeastern
268                             Columbia University  ...  Northeastern

[269 rows x 5 columns]

Notice that the merge automatically occurred on "School Name", which is the only column name shared by the two dataframes.

This could have been more explicitly coded by using the `on` parameter as follows.
* Note that if `on` is not used, the merge defaults to merging on columns on the intersection of the columns of the two dataframes

In [98]:
pd.merge(schools, regions, on = "School Name")

School Name  ...        Region
0                University of California, Berkeley  ...    California
1                      University of Virginia (UVA)  ...      Southern
2                          Cal Poly San Luis Obispo  ...    California
3    University of California at Los Angeles (UCLA)  ...    California
4        University of California, San Diego (UCSD)  ...    California
..                                              ...  ...           ...
264                              Harvard University  ...  Northeastern
265                      University of Pennsylvania  ...  Northeastern
266                              Cornell University  ...  Northeastern
267                                Brown University  ...  Northeastern
268                             Columbia University  ...  Northeastern

[269 rows x 5 columns]

What do we do if the columns that we want to merge on are labeled differently in the two dataframes? Stay tuned...

## The `left_on` and `right_on` Parameters

You will oftentimes come across dataframes that you want to merge, but do not have a key column of the same name on which to merge. We can still do it.

Let's bring in some new data describing the distribution (percentiles) of student median income by school.

In [99]:
income_url = 'https://andybek.com/pandas-mid'

In [100]:
mid_career = pd.read_csv(income_url)

In [101]:
mid_career.head()

school_name  ... Mid-Career 90th Percentile Salary
0   Massachusetts Institute of Technology (MIT)  ...                      $220,000.00 
1      California Institute of Technology (CIT)  ...                               NaN
2                           Harvey Mudd College  ...                               NaN
3  Polytechnic University of New York, Brooklyn  ...                      $190,000.00 
4                                  Cooper Union  ...                               NaN

[5 rows x 5 columns]

Now we want to add this to our *schools* dataframe.

In [102]:
schools.head()

School Name  ... Mid-Career Median Salary
0              University of California, Berkeley  ...             $112,000.00 
1                    University of Virginia (UVA)  ...             $103,000.00 
2                        Cal Poly San Luis Obispo  ...             $101,000.00 
3  University of California at Los Angeles (UCLA)  ...             $101,000.00 
4      University of California, San Diego (UCSD)  ...             $101,000.00 

[5 rows x 4 columns]

But we have a problem - the key columns containing the school names are labeled a bit differently. Let's try `pd.merge()` anyway.

In [103]:
## This will not work because Pandas does not have an identical key to merge on.
# pd.merge(schools, mid_career)

To resolve this issue, we use the `left_on` and `right_on` parameters, where we provide the names of the columns in the "left" and "right" dataframes on which to merge. This allows us to identify exactly which column to merge on, even if they are named differently.

In [104]:
pd.merge(schools, mid_career, left_on = "School Name", right_on = "school_name")

School Name  ... Mid-Career 90th Percentile Salary
0                University of California, Berkeley  ...                      $201,000.00 
1                      University of Virginia (UVA)  ...                      $215,000.00 
2                          Cal Poly San Luis Obispo  ...                      $178,000.00 
3    University of California at Los Angeles (UCLA)  ...                      $193,000.00 
4        University of California, San Diego (UCSD)  ...                      $177,000.00 
..                                              ...  ...                               ...
264                              Harvard University  ...                      $288,000.00 
265                      University of Pennsylvania  ...                      $282,000.00 
266                              Cornell University  ...                      $210,000.00 
267                                Brown University  ...                      $228,000.00 
268                             Columbia University  ...                      $241,000.00 

[269 rows x 9 columns]

This is *almost* what we wanted. The merged dataframe contains all of the data from both constituent dataframes. However, we have two columns with school names which is a bit redundant. There are multiple ways to remove this, but the easiest based on what we already know is to use the `drop()` method.

In [105]:
pd.merge(schools, mid_career, left_on = "School Name", right_on = "school_name").drop(columns = "school_name")

School Name  ... Mid-Career 90th Percentile Salary
0                University of California, Berkeley  ...                      $201,000.00 
1                      University of Virginia (UVA)  ...                      $215,000.00 
2                          Cal Poly San Luis Obispo  ...                      $178,000.00 
3    University of California at Los Angeles (UCLA)  ...                      $193,000.00 
4        University of California, San Diego (UCSD)  ...                      $177,000.00 
..                                              ...  ...                               ...
264                              Harvard University  ...                      $288,000.00 
265                      University of Pennsylvania  ...                      $282,000.00 
266                              Cornell University  ...                      $210,000.00 
267                                Brown University  ...                      $228,000.00 
268                             Columbia University  ...                      $241,000.00 

[269 rows x 8 columns]

## Inner vs Outer Joins

We've learned how to merge dataframes on a *key column*. But we haven't yet discussed how the specific keys (entries) in the dataset are included in the final merge.

An **inner** join only includes common keys in both dataframes. If a key appears in the key column of one dataframe and not the other, it will not appear in the merge. It is similar to a *set intersection*.

An **outer** join will include all keys from both key columns of the two dataframes, regardless of whether the keys appear in one key column or both. It is similar to a *set union*. Any key that does not have a counterpart in the other dataframe will have `NaN` as its value in the merged dataframe. 




The type of join that occurs with the `pd.merge()` method is controlled by `how` parameter

Suppose we want to merge the *ivies* dataframe with the *regions* dataframe. Let's start with the basic merge.

In [106]:
pd.merge(ivies, regions)

School Name  ...        Region
0           Dartmouth College  ...  Northeastern
1        Princeton University  ...  Northeastern
2             Yale University  ...  Northeastern
3          Harvard University  ...  Northeastern
4  University of Pennsylvania  ...  Northeastern
5          Cornell University  ...  Northeastern
6            Brown University  ...  Northeastern
7         Columbia University  ...  Northeastern

[8 rows x 5 columns]

Looks good, all of the Ivy League schools are in the Northeastern U.S. But wait a second - the *regions* dataframe has many schools in it (269 to be exact), and yet the merged dataframe only has the ivies. What gives?

The reason is that `pd.merge()` defaults to an *inner* merge (`how = "inner"`). Only the Ivy League school names were common to the key column *School Name* in each dataframe, and thus only those names were kept.

Let's now try an *outer* merge.

In [108]:
pd.merge(ivies, regions, how='outer')

School Name  ...        Region
0                      Dartmouth College  ...  Northeastern
1                   Princeton University  ...  Northeastern
2                        Yale University  ...  Northeastern
3                     Harvard University  ...  Northeastern
4             University of Pennsylvania  ...  Northeastern
..                                   ...  ...           ...
264         Austin Peay State University  ...      Southern
265           Pittsburg State University  ...    Midwestern
266             Southern Utah University  ...       Western
267  Montana State University - Billings  ...       Western
268         Black Hills State University  ...    Midwestern

[269 rows x 5 columns]

We see here that we have all of the Ivy League schools with Region populated, but no values for School Type or salary. That's because only the Ivy League keys were included in the key column (School Name) of the *ivies* dataframe. The remaining keys (schools) areo nly found in the *regions* dataframe, and thus they will have none of the School Type and Salary data from *ivies* - those values will default to `NaN`.

## Left vs. Right Joins

The `how` parameter of the `merge()` method also supports `left` and `right` joins. This essentially tells Pandas from which dataframe to preserve the keys. For example, if we do a "left" join, only the keys from the *ivies* dataframe will be preserved in the merge.

In [109]:
pd.merge(ivies, regions, how='left')

School Name  ...        Region
0           Dartmouth College  ...  Northeastern
1        Princeton University  ...  Northeastern
2             Yale University  ...  Northeastern
3          Harvard University  ...  Northeastern
4  University of Pennsylvania  ...  Northeastern
5          Cornell University  ...  Northeastern
6            Brown University  ...  Northeastern
7         Columbia University  ...  Northeastern

[8 rows x 5 columns]

If we do a "right" join, the keys from the *regions* dataframe will be preserved. Of course, the *regions* dataframe includes the Ivy League schools, and so this is essentially an outer join.

In [112]:
pd.merge(ivies, regions, how='right')

School Name  ...        Region
0     Massachusetts Institute of Technology (MIT)  ...  Northeastern
1        California Institute of Technology (CIT)  ...    California
2                             Harvey Mudd College  ...    California
3    Polytechnic University of New York, Brooklyn  ...  Northeastern
4                                    Cooper Union  ...  Northeastern
..                                            ...  ...           ...
264                  Austin Peay State University  ...      Southern
265                    Pittsburg State University  ...    Midwestern
266                      Southern Utah University  ...       Western
267           Montana State University - Billings  ...       Western
268                  Black Hills State University  ...    Midwestern

[269 rows x 5 columns]

A final note is that if we flip the order in which we declare the dataframes, the joins will flip as well. For instance, these two code blocks are equivalent:

`pd.merge(ivies, regions, how='left')`

`pd.merge(regions, ivies, how='right')`
